File: research/items/unified-hybrid-software-engine/tests/run_guard.py
# SPDX-License-Identifier: Apache-2.0
"""Keep executable tooling from changing captured evidence."""
import os
from pathlib import Path


def resolve_run(base):
    return Path(os.environ.get('UNIFIED_RUN') or (Path(base) / 'active-run.txt').read_text().strip())


def require_writable_run(run):
    run = Path(run).resolve()
    evidence = Path(__file__).resolve().parent.parent / 'evidence'
    if run.parent != evidence or not run.is_dir():
        raise ValueError('UNIFIED_RUN must name an existing direct child of this item\'s evidence directory')
    if (run / 'manifest.json').exists():
        raise ValueError(f'Sealed evidence is read-only: {run}. Create a new run and set UNIFIED_RUN.')
    return run

File: research/items/unified-hybrid-software-engine/evidence/20260921T192753Z-review-02/tooling/run-guard.mjs
// SPDX-License-Identifier: Apache-2.0
import {access,stat,realpath} from 'node:fs/promises';
import path from 'node:path';
import {fileURLToPath} from 'node:url';

export async function requireWritableRun(run){
 const target=await realpath(run),evidence=fileURLToPath(new URL('../evidence/',import.meta.url));
 if(path.dirname(target)!==path.resolve(evidence)||!(await stat(target)).isDirectory())throw Error('UNIFIED_RUN must name an existing run in this item\'s evidence directory');
 let sealed=true;
 try{await access(path.join(target,'manifest.json'));}catch(error){if(error.code!=='ENOENT')throw error;sealed=false;}
 if(sealed)throw Error(`Sealed evidence is read-only: ${target}. Create a new run and set UNIFIED_RUN.`);
}
